sourced/ml/algorithms/id_splitter/nn_model.py

```diff
@@ -0,0 +1,211 @@
+    return model
+
+
+def add_conv(X, filters=[64, 32, 16, 8], kernel_sizes=[2, 4, 8, 16], output_n_filters=32):
+    """
+    Build a single convolutional layer.
+    :param X: previous layer
+    :param filters: number of filter for each kernel size
+    :param kernel_sizes: list of kernel sizes
+    :param output_n_filters: number of 1d output filters
+    :return: layer
+    """
+    # normalization of input
+    X = BatchNormalization()(X)
+
+    # add convolutions
+    convs = []
+
+    for n_filters, kern_size in zip(filters, kernel_sizes):
+        conv = Conv1D(filters=n_filters, kernel_size=kern_size, padding="same",
+                      activation="relu")
+        convs.append(conv(X))
+
+    # concatenate all convolutions
+    conc = Concatenate(axis=-1)(convs)
+    conc = BatchNormalization()(conc)
+
+    # dimensionality reduction
+    conv = Conv1D(filters=output_n_filters, kernel_size=1, padding="same", activation="relu")
+    return conv(conc)
+
+
+def build_cnn(n_uniq, maxlen, filters, output_n_filters, stack, kernel_sizes, optimizer, device):
+    """
+    Construct a CNN model according to given arguments.
+    :param n_uniq: number of unique items/characters
+    :param maxlen: maximum length of input sequence
+    :param filters: list with number of filters for each kernel size
+    :param output_n_filters: number of 1-d filters that is applied after each CNN layer
+    :param stack: number of CNN layers to stack
+    :param kernel_sizes: list of kernel sizes
+    :param optimizer: optimizer
+    :param device: device to use
+    :return: compiled model
+    """
+    # prepare model
+    with tf.device(device):
+        char_seq, hid_layer = prepare_input_emb(maxlen, n_uniq)
+
+        # stack CNN
+        for _ in range(stack):
+            hid_layer = add_conv(hid_layer, filters=filters, kernel_sizes=kernel_sizes,
+                                 output_n_filters=output_n_filters)
+
+        output = add_output_layer(hid_layer)
+
+    # compile model
+    model = Model(inputs=char_seq, outputs=output)
+
+    model.compile(optimizer=optimizer, loss=LOSS, metrics=METRICS)
+    return model
+
+
+@register_metric
+def precision(y_true, y_pred):
+    """
+    Precision metric. Only computes a batch-wise average of recall.
+    :param y_true: tensor
+    :param y_pred: tensor
+    :return: tensor
+    """
+    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
+    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
+    precision = true_positives / (predicted_positives + K.epsilon())
+    return precision
+
+
+@register_metric
+def recall(y_true, y_pred):
+    """
+    Recall metric. Only computes a batch-wise average of recall.
+    :param y_true: tensor
+    :param y_pred: tensor
+    :return: tensor
+    """
+    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
+    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
+    recall = true_positives / (possible_positives + K.epsilon())
```

Comment:
same here, no need to create recall variable, also `possible_positives` could be refactored to `positives`, or `actual_positives` ? i find it kinda misguiding at first glance, as possible kinda infers lack of certainty, which there is not in this case since we know they are positives

Reply:
https://en.wikipedia.org/wiki/Precision_and_recall
`possible_positives` - it could be `relevant_positive` / `retrieved_positive` if we want to follow wiki naming.
I don't mind renaming - let's see which option is more preferable by other reviewers.